crates/glue/writing_a_glue_plugin.md

```diff
@@ -0,0 +1,441 @@
+    uint8_t red;
+};
+```
+
+### HasClosure
+
+(TODO: no glue spec currently supports this. Unsure if this either does not work or if nobody who knows how has gotten around to it.)
+
+## TagUnion
+
+Roc's tagged union/sum types. Roc may choose to optimize a tag union's in-memory representation behind the scenes. This results in tag unions coming in seven different flavors.
+
+Some info common across all of these are:
+- All of these are guaranteed have at least one tag (see also: [EmptyTagUnion](#emptytagunion)).
+- Discriminants (when stored in a discriminant field or as tags in a tagged pointer) are equal to a tag's index in the list the variants are given in (e.g: the `tags` field in an `Enumeration`) (if applicable).
+- Discriminants (when stored in a discriminant field) are currently either an 8-bit or 16-bit unsigned integer.
+
+### NonRecursive
+
+The base case: a tag union with more than one tag which does not contain itself, even through indirection.
+
+```roc
+# in `platform.roc`
+Cell : [
+    Empty,
+    Number I32,
+    Text Str,
+]
+```
+```roc
+# shape reported by `roc glue`
+TagUnion (NonRecursive {
+    name: "Cell",
+    discriminantSize: 1,
+    discriminantOffset: 24,
+    tags: [
+        {name: "Empty", payload: None},
+        {name: "Number", payload: (Some (@TypeId 1))},
+        {name: "Text", payload: (Some (@TypeId 2))}
+    ]
+})
+```
+```c
+// example C layout
+// note: it is currently guaranteed that the discriminant comes after the
+// space reserved in memory for the union, including padding. this means
+// NonRecursive tag unions may be represented as a struct, as below. this may
+// change in the future, as this currently makes for a missed optimization:
+// https://github.com/roc-lang/roc/issues/7171
+struct Cell {
+    union {
+        int32_t Number;
+        RocStr Text;
+    } payload;
+    enum {
+        Empty, Number, Text
+    } discriminant;
+};
+```
+
+### Enumeration
+
+A special case where a none of the tags in a tag union contains a payload. This means it can be represented as just its discriminant.
+
+```roc
+# in `platform.roc`
+Color : [Red, Green, Blue]
+```
+```roc
+# shape reported by `roc glue`
+TagUnion (Enumeration {
+    name: "Expr", 
+    size: 1,
+    tags: ["Blue", "Green", "Red"]
+})
+```
+```c
+// example C layout
+enum Color {
+    Blue = 0,
+    Green = 1,
+    Red = 2,
+};
+```
+
+### SingleTagStruct
+
+A special case where a tag union contains exactly one tag. This means it can be represented as just that tag's payload.
+
+(TODO: the only difference between a `SingleTagStruct` and a `NonNullableUnwrapped` is that the former is non-recursive. It seems this does not result in any differences in layout. Is this one of these shapes vestigial?)
+
+### Recursive
+
+A recursive tag union ineligable for other optimizations.
+
+(TOOO: the `Recursive` variant should only occur when all tags have payloads. Couldn't it store a `TypeId` per payload instead of a `[None, Some TypeId]`?)
```

Comment:
Yeah, sounds like an oversight due to `NullableWrapped` being an optimization that was added later. We definitely could use some general naming cleanup.